src/service.rs

```diff
@@ -1,56 +1,74 @@
 //! HTTP gateway service for Holochain
 
-use std::{
-    net::{IpAddr, SocketAddr},
-    sync::Arc,
-};
+use std::net::{IpAddr, SocketAddr};
 
-use axum::{routing::get, Router};
+use axum::{middleware, routing::get, Router};
 use tokio::net::TcpListener;
 
-use crate::{config::Configuration, error::HcHttpGatewayResult, routes::health_check};
+use crate::{
+    config::Configuration,
+    error::HcHttpGatewayResult,
+    middleware::validate_zome_call_payload_size,
+    routes::{health_check, zome_call},
+    HcHttpGatewayError,
+};
 
 /// Core Holochain HTTP gateway service
 #[derive(Debug)]
 pub struct HcHttpGatewayService {
-    address: SocketAddr,
+    listener: TcpListener,
     router: Router,
 }
 
 /// Shared application state
 #[derive(Debug, Clone)]
 pub struct AppState {
-    configuration: Configuration,
+    pub configuration: Configuration,
 }
 
 impl HcHttpGatewayService {
     /// Create a new service instance bound to the given address and port
-    pub fn new(address: impl Into<IpAddr>, port: u16, configuration: Configuration) -> Self {
+    pub async fn new(
+        address: impl Into<IpAddr>,
+        port: u16,
+        configuration: Configuration,
+    ) -> HcHttpGatewayResult<Self> {
         let address = SocketAddr::new(address.into(), port);
 
-        let state = Arc::new(AppState { configuration });
+        let state = AppState { configuration };
 
         let router = Router::new()
+            .route(
+                "/{dna_hash}/{coordinator_identifier}/{zome_name}/{function_name}",
+                get(zome_call),
+            )
             .route("/health", get(health_check))
+            .layer(middleware::from_fn_with_state(
+                state.clone(),
+                validate_zome_call_payload_size,
+            ))
             .with_state(state.clone());
 
         tracing::info!("Configuration: {:?}", state.configuration);
 
-        HcHttpGatewayService { router, address }
+        let listener = TcpListener::bind(address).await?;
+
+        Ok(HcHttpGatewayService { router, listener })
     }
 
     /// Get the socket address the service is configured to use
-    pub fn address(&self) -> SocketAddr {
-        self.address
+    pub fn address(&self) -> HcHttpGatewayResult<SocketAddr> {
```

Comment:
Nice, this keeps the listener encapsulated and achieves the same result as the axum testing example. I like it